UOP1_Project/Assets/Scripts/Characters/WalkingParticle.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class WalkingParticle : MonoBehaviour
+{
+
+    [Tooltip("Particle prefab to be attached to the charactor in the scene.")] public ParticleSystem walkingParticlePrefab;
+
+    private Vector3 lastPosition;
+    private Vector3 currentPosition;
+    private ParticleSystem particleObject;
+
+    private void Update()
+    {
+        if (lastPosition == null) lastPosition = transform.position;
+        if (currentPosition == null) currentPosition = transform.position;
+        if (particleObject == null && walkingParticlePrefab != null)
+        {
+            particleObject = Instantiate(walkingParticlePrefab, transform);
+            particleObject.Stop();
+        }
+        if(currentPosition != null && lastPosition != null && particleObject != null)
+        {
+            currentPosition = transform.position;
+            if(currentPosition.x * 1000 != lastPosition.x * 1000 || currentPosition.z * 1000 != lastPosition.z * 1000)
```

Comment:
better define a treshold to check against instead of the multiplication, e.g.

`if (Mathf.Abs(currentPosition.x) < 0.001)` etc. Define treshold as a const somewhere in the project.

Reply:
Even simpler: use [Vector2.Distance()](https://docs.unity3d.com/ScriptReference/Vector2.Distance.html) to calculate the distance between lastPosition and currentPosition. Then you don't need to do separate checks for x/z.

If the X/Z position is all that ever matters here - never the Y - then currentPosition/lastPosition could be changed to be Vector2 fields as well, and then you'd assign them like `currentPosition = new Vector2(transform.position.x, transform.position.z)`.